pkg/reconciler/inmemorychannel/dispatcher/inmemorychannel.go

```diff
@@ -19,87 +19,96 @@ package dispatcher
-	}
-	err = r.dispatcher.UpdateConfig(ctx, r.eventDispatcherConfigStore.GetConfig(), config)
-	if err != nil {
-		logging.FromContext(ctx).Error("Error updating InMemory dispatcher config")
+		logging.FromContext(ctx).Error("Error creating config for in memory channels", zap.Error(err))
 		return err
 	}
 
+	// First grab the MultiChannelFanoutMessage handler
+	handler := r.multiChannelMessageHandler.GetChannelHandler(config.HostName)
+	if handler == nil {
+		// No handler yet, create one.
+		fanoutHandler, err := fanout.NewFanoutMessageHandler(logging.FromContext(ctx).Desugar(), channel.NewMessageDispatcher(logging.FromContext(ctx).Desugar()), config.FanoutConfig, r.reporter)
+		if err != nil {
+			logging.FromContext(ctx).Error("Failed to create a new fanout.MessageHandler", err)
+			return err
+		}
+		r.multiChannelMessageHandler.SetChannelHandler(config.HostName, fanoutHandler)
+	} else {
+		// Just update the config if necessary.
+		haveSubs := handler.GetSubscriptions(ctx)
+		if diff := cmp.Diff(config.FanoutConfig.Subscriptions, haveSubs, cmpopts.IgnoreFields(kncloudevents.RetryConfig{}, "Backoff", "CheckRetry")); diff != "" {
+			logging.FromContext(ctx).Info("Updating fanout config: ", zap.String("Diff", diff))
+			handler.SetSubscriptions(ctx, config.FanoutConfig.Subscriptions)
+		}
+	}
 	return nil
 }
 
-// newConfigFromInMemoryChannels creates a new Config from the list of inmemory channels.
-func (r *Reconciler) newConfigFromInMemoryChannels(channels []*v1.InMemoryChannel) (*multichannelfanout.Config, error) {
-	cc := make([]multichannelfanout.ChannelConfig, 0)
-	for _, c := range channels {
+func (r *Reconciler) FinalizeKind(ctx context.Context, imc *v1.InMemoryChannel) reconciler.Event {
+	if imc.Status.Address != nil &&
+		imc.Status.Address.URL != nil {
+		if hostName := imc.Status.Address.URL.Host; hostName != "" {
+			logging.FromContext(ctx).Info("Removing dispatcher")
+			r.multiChannelMessageHandler.DeleteChannelHandler(hostName)
+		}
+	}
+	return nil
+
+}
 
-		subs := make([]fanout.Subscription, len(c.Spec.Subscribers))
-		for _, sub := range c.Spec.Subscribers {
-			conf, err := fanout.SubscriberSpecToFanoutConfig(sub)
-			if err != nil {
-				return nil, err
-			}
+// newConfigForInMemoryChannel creates a new Config for a single inmemory channel.
+func (r *Reconciler) newConfigForInMemoryChannel(imc *v1.InMemoryChannel) (*multichannelfanout.ChannelConfig, error) {
+	subs := make([]fanout.Subscription, len(imc.Spec.Subscribers))
 
-			subs = append(subs, *conf)
+	for i, sub := range imc.Spec.Subscribers {
+		conf, err := fanout.SubscriberSpecToFanoutConfig(sub)
+		if err != nil {
+			return nil, err
 		}
-
-		channelConfig := multichannelfanout.ChannelConfig{
-			Namespace: c.Namespace,
-			Name:      c.Name,
-			HostName:  c.Status.Address.URL.Host,
-			FanoutConfig: fanout.Config{
-				AsyncHandler:  true,
-				Subscriptions: subs,
-			},
+		if conf.DeadLetter == nil && conf.Reply == nil && conf.Subscriber == nil {
```

Comment:
Does this ever happen? Isn't a sub like that invalid and should be catched before by some other controller/webhook?

Reply:
Yes, this is left over cruft. Removing...